src/odemis/driver/keysight.py

```diff
@@ -0,0 +1,333 @@
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+"""
+
+import logging
+import re
+import time
+import socket
+
+from odemis import model
+from odemis.model import HwError
+from odemis.util import to_str_escape
+from typing import List, Tuple
+
+
+class TrueForm(model.Emitter):
+    def __init__(self, name, role, address, channels: List[int],
+                 limits: List[Tuple[float, float]], **kwargs):
+        """ Initializes the Keysight 33600 series Arbitrary Waveform Generator.
+        :param name: (str) as in Odemis
+        :param role: (str) as in Odemis
+        :param address: # “fake” (to start the simulator) or an IP address
+        :param channels: # Channels which generate the waveform
+        :param limits: # min/max V for each channel. if min>max, the polarity of the channel is inverted.
+        """
+        delay = "DEFAULT VALUE"
+        self.delay = model.FloatContinuous(delay, setter=self._setTimeRange, unit="s")
+
+        dutyCycle = "DEFAULT VALUE"
+        self.dutyCycle = model.FloatContinuous(dutyCycle, setter=self.setDutyCycle, unit="s")
+
+        period = 4e-6  # default value
+        self.period = model.FloatContinuous(period, setter=self.setPeriod, unit="s")
+
+        power = "DEFAULT VALUE"
+        self.power = model.BooleanVA(power, setter=self._setTimeRange, unit="s")
+
+    def _findDevice(self, address):
+        """
+        Look for a compatible device
+        address (None or int): the address of the Waveform Generator
+        return (serial, int): the (opened) serial port used, and the actual address
+        raises:
+            IOError: if no device are found
+        """
+        # Connection via ethernet cable
+        if not address.startswith("/"):
+            try:
+                self._accesser = IPBusAccesser(address)
+            except Exception as e:
+                logging.info("Could not establish connection to the device through IP bus accesser: %s", e)
+                raise
+
+            return address
+
+        if address == "/dev/fake":
+            return address
+
+        else:
+            raise HwError("Failed to find a device on the address: '%s'. "
+                          "Check that the device is turned on and connected to "
+                          "the computer." % (address,))
+
+    def _sendCommand(self, cmd):
+        """
+        cmd (str): command to be sent to device
+        """
+        # Send the given command to the instrument
+        logging.debug("Sending command %s", to_str_escape(cmd))
+        self._accesser.sendCmd(cmd)  # TODO: check with real hw if a brief wait needed(like 50 ms of time.sleep)
+
+    def _sendQuery(self, q):
```

Comment:
The point of sendQuery, compared to sendCommand is that it *returns* something. => Indicate it in the signature, and actually return the value.